x-pack/libbeat/reader/etw/syscall_advapi32.go

```diff
@@ -0,0 +1,347 @@
+	dwLowDateTime  uint32
+	dwHighDateTime uint32
+}
+
+// https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-systemtime
+type SystemTime struct {
+	Year         uint16
+	Month        uint16
+	DayOfWeek    uint16
+	Day          uint16
+	Hour         uint16
+	Minute       uint16
+	Second       uint16
+	Milliseconds uint16
+}
+
+// https://learn.microsoft.com/en-us/windows/win32/api/evntrace/nf-evntrace-enabletrace
+const (
+	TRACE_LEVEL_NONE        = 0
+	TRACE_LEVEL_CRITICAL    = 1
+	TRACE_LEVEL_FATAL       = 1
+	TRACE_LEVEL_ERROR       = 2
+	TRACE_LEVEL_WARNING     = 3
+	TRACE_LEVEL_INFORMATION = 4
+	TRACE_LEVEL_VERBOSE     = 5
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/api/evntprov/nc-evntprov-penablecallback
+const (
+	EVENT_CONTROL_CODE_DISABLE_PROVIDER = 0
+	EVENT_CONTROL_CODE_ENABLE_PROVIDER  = 1
+	EVENT_CONTROL_CODE_CAPTURE_STATE    = 2
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/api/evntrace/nf-evntrace-controltracea
+const (
+	EVENT_TRACE_CONTROL_QUERY  = 0
+	EVENT_TRACE_CONTROL_STOP   = 1
+	EVENT_TRACE_CONTROL_UPDATE = 2
+	EVENT_TRACE_CONTROL_FLUSH  = 3
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/api/evntrace/ns-evntrace-event_trace_logfilea
+const (
+	PROCESS_TRACE_MODE_REAL_TIME     = 0x00000100
+	PROCESS_TRACE_MODE_RAW_TIMESTAMP = 0x00001000
+	PROCESS_TRACE_MODE_EVENT_RECORD  = 0x10000000
+)
+
+const INVALID_PROCESSTRACE_HANDLE = 0xFFFFFFFFFFFFFFFF
+
+// https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes
+const (
+	ERROR_ACCESS_DENIED          syscall.Errno = 5
+	ERROR_INVALID_HANDLE         syscall.Errno = 6
+	ERROR_BAD_LENGTH             syscall.Errno = 24
+	ERROR_INVALID_PARAMETER      syscall.Errno = 87
+	ERROR_INSUFFICIENT_BUFFER    syscall.Errno = 122
+	ERROR_BAD_PATHNAME           syscall.Errno = 161
+	ERROR_ALREADY_EXISTS         syscall.Errno = 183
+	ERROR_NOT_FOUND              syscall.Errno = 1168
+	ERROR_NO_SYSTEM_RESOURCES    syscall.Errno = 1450
+	ERROR_TIMEOUT                syscall.Errno = 1460
+	ERROR_WMI_INSTANCE_NOT_FOUND syscall.Errno = 4201
+	ERROR_CTX_CLOSE_PENDING      syscall.Errno = 7007
+	ERROR_EVT_INVALID_EVENT_DATA syscall.Errno = 15005
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/etw/logging-mode-constants (to extend modes)
+// https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/wmistr/ns-wmistr-_wnode_header (to extend flags)
+const (
+	WNODE_FLAG_ALL_DATA        = 0x00000001
+	WNODE_FLAG_TRACED_GUID     = 0x00020000
+	EVENT_TRACE_REAL_TIME_MODE = 0x00000100
+)
+
+type GUID struct {
+	Data1 uint32
+	Data2 uint16
+	Data3 uint16
+	Data4 [8]byte
+}
+
+// For testing purposes we create a variable to store the function to call
+// When running tests, these variables point to a mock function
+var (
+	StartTraceFunc   = _StartTrace
+	EnableTraceFunc  = _EnableTraceEx2
+	ControlTraceFunc = _ControlTrace
+	OpenTraceFunc    = _OpenTrace
+	ProcessTraceFunc = _ProcessTrace
+	CloseTraceFunc   = _CloseTrace
+)
```

Comment:
This approach to test mocking makes me feel queasy. If we have fields that hold `func` values for these functions on `Session` and methods to call them that fall to the functions here if nil, and otherwise call the function in the `func` value, then mocking can be done on a per value basis with no risk of concurrent mutation in the future and no need to do the retain old/mutate/restore dance.

Reply:
Not sure if I understand this feedback. My original problem is having to mock functions that call Windows API, so my approach is to create an alias for these functions that point to the original function in production code, and then in test functions they point to a mock.

`Session` (if you mean the struct) only stores pointers to the callbacks because I need them to point to the callback I created in the Filebeat input. It doesn't store any pointer to these functions that I need to mock.

Reply:
For example [here](https://github.com/elastic/beats/pull/36914/files#diff-8bc822f6e6a1a389837b7572bb8fe6b4606b5671f7e5a425083012c3a11928eaR16-R26). This could be avoided if `Session` (the struct) had a `ControlTraceFunc` field that could be called instead of using a global.

Reply:
Thanks for the clarifications Dan! I've done the requested changes in 3549a10abf7da40d20c509110491b7ca53d79262. However, there are still a couple of mock functions that cannot be added to the `Session` struct since they are called before an instance of `Session` is created.